# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [5]:
def addTwo(x):
    return x+2
print(addTwo(4))
#lambda version
#without using a var to pass in data
print((lambda x:x+2)(4))
#using a var
a_num = 4
print((lambda x: x+2)(4))

6
6
6


#### Saving to a Variable

In [6]:
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [10]:
#multiple input with no vars
print((lambda x, y, z: x*y*z)(3,5,8))
#multiple inputs with var
x_test = lambda x, y ,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [ ]:
def multiply(f,num):
    '''
        f expects lambda fcntn
        num expects number
    '''
    return f(num)
multiply(lambda x: x*x, 4)

#### Returning a Lambda from a Function

In [11]:
#regular defined function
def multiply_test(num):
    return num*4
#function within a function
def returnFunc():
    test =4 
    def multiply(num):
        return num*2
    return multiply
f_return = returnFunc()
print(returnFunc)
print(f_return(4))

<function returnFunc at 0x7fe1c057ab80>
8


#### If Statements within Lambdas

In [13]:
#lambda x: true if cond else false
f_condition = lambda num: num*2 if num> 10 else num +2
print(f_condition(8))

10


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
x = lambda f: f**3
x(6)

216

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
def squared(a,b):
    if a < 10 and b < 10:
        return a**2, b**2
    else:
        return a, b
numbers = [1,3,5,2,342,235,12,123,122,35,6,2]
num2 = [1,2,3,4,5,6,7,8,8,9,5678]
squaremap = map(squared, numbers, num2)
print(list(squaremap))
 

[(1, 1), (9, 4), (25, 9), (4, 16), (342, 5), (235, 6), (12, 7), (123, 8), (122, 8), (35, 9), (6, 5678)]


#### Using Lambda's with Map

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [23]:
numbers = [1,3,5,2,342,235,12,123,122,35,6,2]
mapp = map(lambda x: 2*x-1, numbers)
print(list(mapp))

[1, 5, 9, 3, 683, 469, 23, 245, 243, 69, 11, 3]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [25]:
names = ['Bob','Andy','Max','Evan','Lolo']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy']


#### Using Lambda's with Filter()

In [93]:
names = ['Bob','Andy','Max','Evan','Lolo']
new_names_lambda = list(filter(lambda name: True if name[0].lower == 'a' else False, names))
print(new_names_lambda)

TypeError: 'list' object is not callable

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [107]:
import statistics
listt = [1,2,3,5,1,124,123,23,2,2,3,234,25,62,133,12,23,73,234,12,32,2,2]
m = statistics.mean(listt)
print(m)
f = set(filter(lambda x: True if x >= m else False, listt))
print(f)


49.26086956521739
{133, 73, 234, 123, 124, 62}


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

addnums(5) = 5 +addnums(4)


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [108]:
listt = [1,2,5,2,3,23,6,2,6,2]
a = 1
for x in listt:
    a += a*x
print(a)

4572288


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [117]:
def addnums(num):
    #base case
    if num <=1:
        return num
    else:
        return num + addnums(num-1)
addnums(1000)

500500

#### Writing a Factorial Function

In [119]:
def fac(num):
    if num <=1:
        return num
    else:
        return num*fac(num-1)
print(fac(69))

171122452428141311372468338881272839092270544893520369393648040923257279754140647424000000000000000


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

5


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [131]:
def r(stop, start, step = -1):
    while stop <=start:
        yield start
        start += step
for i in r(0,start = 10):
    a = i**2
    print(a)
    

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
a = set(filter(lambda x: False if x.strip() == "" else True, places))
places = []
for x in a:
    places.append(x)
print(places)



### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key=lambda x: x.split()[-1].lower())
print(author)



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [153]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
def c(x):
    return (x[0],9*x[1]/5+32)
places = set(map(c,places))
print(places)


{('Nashua', 89.6), ('Los Angelos', 111.2), ('Miami', 84.2), ('Boston', 53.6)}


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
nt = int(input("How many terms do you want?"))
def a(n):
    if n <= 1:
        return n
    else:
        return a(n-1)+a(n-2)
for i in range(nt):
    print("Iteration",i,": ",a(i+1))
        

How many terms do you want?99
Iteration 0 :  1
Iteration 1 :  1
Iteration 2 :  2
Iteration 3 :  3
Iteration 4 :  5
Iteration 5 :  8
Iteration 6 :  13
Iteration 7 :  21
Iteration 8 :  34
Iteration 9 :  55
Iteration 10 :  89
Iteration 11 :  144
Iteration 12 :  233
Iteration 13 :  377
Iteration 14 :  610
Iteration 15 :  987
Iteration 16 :  1597
Iteration 17 :  2584
Iteration 18 :  4181
Iteration 19 :  6765
Iteration 20 :  10946
Iteration 21 :  17711
Iteration 22 :  28657
Iteration 23 :  46368
Iteration 24 :  75025
Iteration 25 :  121393
Iteration 26 :  196418
Iteration 27 :  317811
Iteration 28 :  514229
Iteration 29 :  832040
Iteration 30 :  1346269
Iteration 31 :  2178309
Iteration 32 :  3524578
Iteration 33 :  5702887
Iteration 34 :  9227465
Iteration 35 :  14930352
Iteration 36 :  24157817
Iteration 37 :  39088169
Iteration 38 :  63245986
Iteration 39 :  102334155
Iteration 40 :  165580141
Iteration 41 :  267914296
Iteration 42 :  433494437
Iteration 43 :  701408733
Iteration 44 :  11